Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Internal/Rating.hs

```diff
@@ -0,0 +1,106 @@
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+
+module Domain.Action.Internal.Rating where
+
+import Data.Foldable ()
+import Data.OpenApi (ToSchema)
+import qualified Domain.Types.Person as DP
+import qualified Domain.Types.Rating as DRating
+import Domain.Types.Ride (RideStatus (..))
+import qualified Domain.Types.Ride as DRide
+import qualified EulerHS.Language as L
+import EulerHS.Prelude hiding (id)
+import Kernel.Beam.Functions as B
+import Kernel.Types.APISuccess (APISuccess (Success))
+import Kernel.Types.CacheFlow
+import Kernel.Types.Error
+import Kernel.Types.Id
+import Kernel.Utils.Common hiding (id)
+import SharedLogic.CallBPPInternal (DriverOfferBppInternal)
+import qualified Storage.Queries.Booking as QBooking
+import qualified Storage.Queries.Person as QP
+import qualified Storage.Queries.Rating as QRating
+import qualified Storage.Queries.Ride as QRide
+import Tools.Error (RatingError (..))
+
+data FeedbackReq = FeedbackReq
+  { rideId :: Id DRide.BPPRide,
+    ratingValue :: Int,
+    feedbackDetails :: Maybe Text
+  }
+  deriving (Show, Generic, ToJSON, FromJSON, ToSchema)
+
+rating ::
+  ( MonadFlow m,
+    EsqDBFlow m r,
+    CacheFlow m r,
+    EncFlow m r,
+    HasFlowEnv m r '["minimumDriverRatesCount" ::: Int],
+    HasFlowEnv m r '["driverOfferBppInternal" ::: DriverOfferBppInternal]
+  ) =>
+  Maybe Text ->
+  FeedbackReq ->
+  m APISuccess
+rating apiKey FeedbackReq {..} = do
+  driverOfferBppInternal <- asks (.driverOfferBppInternal)
+  ride <- B.runInReplica $ QRide.findByBPPRideId rideId >>= fromMaybeM (RideDoesNotExist rideId.getId)
+  booking <- B.runInReplica $ QBooking.findById ride.bookingId >>= fromMaybeM (BookingDoesNotExist ride.bookingId.getId)
+  _ <- validateRequest ratingValue ride.status
+  unless (Just driverOfferBppInternal.internalKey == apiKey) $
+    throwError $ AuthBlocked "Invalid BPP internal api key"
+
+  ratingu <- B.runInReplica $ QRating.findRatingForRide ride.id
+  _ <- case ratingu of
+    Nothing -> do
+      logTagInfo "FeedbackAPI" $
+        "Creating a new record for " +|| ride.id ||+ " with rating " +|| ratingValue ||+ "."
+      newRating <- buildRating ride.id booking.riderId ratingValue feedbackDetails
+      QRating.create newRating
+    Just rideRating -> do
+      logTagInfo "FeedbackAPI" $
+        "Updating existing rating for " +|| ride.id ||+ " with new rating " +|| ratingValue ||+ "."
+      QRating.updateRating rideRating.id booking.riderId ratingValue feedbackDetails
+  calculateAverageRating booking.riderId
+  pure Success
+
+calculateAverageRating ::
+  (EsqDBFlow m r, EncFlow m r, HasFlowEnv m r '["minimumDriverRatesCount" ::: Int]) =>
+  Id DP.Person ->
+  m ()
+calculateAverageRating personId = do
+  logTagInfo "PersonAPI" $ "Recalculating average rating for driver " +|| personId ||+ ""
+  allRatings <- B.runInReplica $ QRating.findAllRatingsForPerson personId
```

Comment:
fetch person current RatingCount and Rating Score.
then Update ratingCount++, ratingScore = givenRating+currRatingScore